### PR TITLE
Add recursive query type support

### PR DESCRIPTION
**Pull request by nihalbhatnagar**
Summary

  Adds support for recursive/self-referential types in query definitions via the new typeReferences map on QueryTypeV2.

Generates types in a non-exported namespaced by their type ID. Expands type for intellisense

### CODE REVIEW

File: packages/generator/src/shared/getInterfaceTypesFromQueryDataType.ts [hunk 1]
```diff
@@ -60,12 +60,10 @@ export function getInterfaceTypesFromQueryDataType(
     case "threeDimensionalAggregation":
     case "timestamp":
     case "twoDimensionalAggregation":
-    case "entrySet":
```
Comment (nihalbhatnagar) [inline packages/generator/src/shared/getInterfaceTypesFromQueryDataType.ts:63]: Duplicates

File: packages/generator/src/v2.0/generatePerQueryDataFiles.ts [hunk 2]
```diff
@@ -127,6 +130,10 @@ async function generateV2QueryFile(
     query.fullApiName,
   );
 
+  const typeRefs = query.raw.typeReferences ?? {};
+  const customTypesNs = generateCustomTypesNamespace(ontology, typeRefs);
+  const typeRefNames = buildTypeRefNames(typeRefs, true);
```
Comment (KuberSethi) [inline packages/generator/src/v2.0/generatePerQueryDataFiles.ts:135]: could we just run this once and then pass the result into `generateCustomTypesNamespace` + reuse?

File: packages/generator/src/v2.0/generatePerQueryDataFiles.ts [hunk 3]
```diff
@@ -390,21 +395,67 @@ export function getQueryParamType(
...
   return paramType;
 }
+
+function buildTypeRefNames(
+  typeRefs: Record<string, QueryDataType>,
+  expand: boolean,
+): Map<string, string> {
+  const names = new Map<string, string>();
+  for (const id of Object.keys(typeRefs)) {
+    const sanitized = `$${id.replace(/-/g, "_")}`;
```
Comment (KuberSethi) [inline packages/generator/src/v2.0/generatePerQueryDataFiles.ts:433]: Here and below in `generateCustomTypesNamespace`, unless this is widely known to be our format can we add a comment explaining why we're using this ID sanitization scheme

File: packages/generator/src/shared/getObjectTypesFromQueryDataType.ts [hunk 4]
```diff
@@ -64,12 +64,10 @@ export function getObjectTypesFromQueryDataType(
     case "threeDimensionalAggregation":
     case "timestamp":
     case "twoDimensionalAggregation":
-    case "entrySet":
+    case "typeReference":
```
Comment (KuberSethi) [inline packages/generator/src/shared/getObjectTypesFromQueryDataType.ts:67]: this makes sense but why the other removals?

File: packages/generator/src/shared/getInterfaceTypesFromQueryDataType.ts [hunk 5]
```diff
@@ -60,12 +60,10 @@ export function getInterfaceTypesFromQueryDataType(
...
+    case "typeReference":
```
Comment (KuberSethi) [inline packages/generator/src/shared/getInterfaceTypesFromQueryDataType.ts:63]: this makes sense but why the other removals?

File: packages/generator-converters/src/wireQueryDataTypeToQueryDataTypeDefinition.ts
Comment (KuberSethi) [inline packages/generator-converters/src/wireQueryDataTypeToQueryDataTypeDefinition.ts]: nit: remove

File: packages/generator/src/v2.0/generatePerQueryDataFiles.ts [hunk 6]
```diff
@@ -390,21 +395,67 @@ export function getQueryParamType(
...
+      paramType = `Partial<Record<${recurse(input.keyType, true)}, ${
+        recurse(input.valueType)
+      }>>`;
+      break;
+
+    case "typeReference":
+      if (!typeRefNames?.has(input.typeId)) {
+        throw new Error(`Unknown typeReference: ${input.typeId}`);
+      }
+      paramType = typeRefNames.get(input.typeId)!;
```
Comment (KuberSethi) [inline packages/generator/src/v2.0/generatePerQueryDataFiles.ts]: do we need the `!` assertion?

File: packages/generator/src/v2.0/generatePerQueryDataFiles.ts [hunk 7]
```diff
@@ -390,21 +395,67 @@ export function getQueryParamType(
...
+function generateCustomTypesNamespace(
+  ontology: EnhancedOntologyDefinition,
+  typeRefs: Record<string, QueryDataType>,
+): string {
+  const names = buildTypeRefNames(typeRefs, false);
+  if (names.size === 0) return "";
+
+  const interfaces: string[] = [];
+  for (const [id, dt] of Object.entries(typeRefs)) {
+    if (dt.type !== "struct") continue;
```
Comment (KuberSethi) [inline packages/generator/src/v2.0/generatePerQueryDataFiles.ts]: could you help me understand why we're skipping structs here but not in `buildTypeRefNames`? Guessing I'm missing something but just want to understand

File: packages/generator/src/v2.0/generatePerQueryDataFiles.test.ts [hunk 8]
```diff
@@ -529,4 +529,587 @@ describe("generatePerQueryDataFiles", () => {
         "
       `);
   });
+
+  it("generates interfaces for recursive type references", async () => {
```
Comment (KuberSethi) [inline packages/generator/src/v2.0/generatePerQueryDataFiles.test.ts:533]: Is it worth adding more tests? These are good but I think given the size/complexity of this change it could benefit from additional testing. Will leave it up to you though!

File: packages/generator-converters/src/wireQueryDataTypeToQueryDataTypeDefinition.ts
Comment (nihalbhatnagar) [reply packages/generator-converters/src/wireQueryDataTypeToQueryDataTypeDefinition.ts]: flop thanks

File: packages/generator/src/shared/getInterfaceTypesFromQueryDataType.ts (hunk 5 above)
Comment (nihalbhatnagar) [reply packages/generator/src/shared/getInterfaceTypesFromQueryDataType.ts:63]: Media reference also a duplicate

File: packages/generator/src/v2.0/generatePerQueryDataFiles.test.ts (hunk 8 above)
Comment (nihalbhatnagar) [reply packages/generator/src/v2.0/generatePerQueryDataFiles.test.ts:533]: Is there a specific edge case you want a test for?  This feels like good coverage

File: packages/generator/src/shared/getObjectTypesFromQueryDataType.ts (hunk 4 above)
Comment (nihalbhatnagar) [reply packages/generator/src/shared/getObjectTypesFromQueryDataType.ts:67]: same as above media reference above

File: packages/generator/src/v2.0/generatePerQueryDataFiles.ts (hunk 2 above)
Comment (nihalbhatnagar) [reply packages/generator/src/v2.0/generatePerQueryDataFiles.ts:135]: No we need both the unexpanded names (because we only want to expand the top layer for intellisense and put the raw reference inside the actual struct

File: packages/generator/src/v2.0/generatePerQueryDataFiles.ts (hunk 3 above)
Comment (nihalbhatnagar) [reply packages/generator/src/v2.0/generatePerQueryDataFiles.ts:433]: Sure I'll add the comment, but IDs start with numbers so we need the prefix

File: packages/generator/src/v2.0/generatePerQueryDataFiles.ts (hunk 7 above)
Comment (nihalbhatnagar) [reply packages/generator/src/v2.0/generatePerQueryDataFiles.ts]: Oversight in buildTypeRefNames, but whats coming back over the wire is always struct. So im just gonna throw if we see a non struct in either place

File: packages/generator/src/v2.0/generatePerQueryDataFiles.ts (hunk 6 above)
Comment (nihalbhatnagar) [reply packages/generator/src/v2.0/generatePerQueryDataFiles.ts]: just made a regular {} since map doesnt have type assertion and we needed it

File: packages/generator/src/v2.0/generatePerQueryDataFiles.test.ts (hunk 8 above)
Comment (KuberSethi) [reply packages/generator/src/v2.0/generatePerQueryDataFiles.test.ts:533]: Think it's fine, although potential improvements could be:
 - non-struct typeRefs
 - different positions of typeRef inside the struct

**Review by KuberSethi — APPROVED** [review]
